File: duplicate-record-write-conflict/python/duplicate_record_write_conflict.py
"""List DNS records before writing, then create or update instead of
blindly POSTing a duplicate. Repairs the write through Cloudflare.

Safe by default. Set DRY_RUN=false to let it write.

Env vars:
  DNS_DOMAIN               the name to check, e.g. "app.example.com"
  DNS_RECORD_TYPE          record type to check, default "A"
  DNS_RECORD_CONTENT       desired content, e.g. an IP for an A record
  DNS_RECORD_TTL           desired TTL, default 300
  DNS_RECORD_PROXIED       "true" or "false", default "false"
  CLOUDFLARE_API_TOKEN     Cloudflare API token (only needed for repair)
  CLOUDFLARE_ZONE_ID       Cloudflare zone id (only needed for repair)
  DRY_RUN                  default "true"; set to "false" to actually write
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("duplicate_record_write_conflict")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "app.example.com")
DNS_RECORD_TYPE = os.environ.get("DNS_RECORD_TYPE", "A")
DNS_RECORD_CONTENT = os.environ.get("DNS_RECORD_CONTENT", "203.0.113.10")
DNS_RECORD_TTL = int(os.environ.get("DNS_RECORD_TTL", "300"))
DNS_RECORD_PROXIED = os.environ.get("DNS_RECORD_PROXIED", "false").lower() == "true"
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

CF_API = "https://api.cloudflare.com/client/v4"


def plan_dns_write(existing_records, desired):
    """Pure decision function. No I/O.

    existing_records: list of {id, name, type, content, ttl, proxied}
                       already at (desired['name'], desired['type'])
    desired: {name, type, content, ttl, proxied}

    Returns one of:
      {"action": "create", "body": desired}
      {"action": "noop", "id": <id>}
      {"action": "update", "id": <id>, "body": {changed fields only}}

    Pure decision logic, no I/O: given zero existing records -> create;
    given one existing record identical to desired -> noop;
    given one existing record differing in content/ttl/proxied -> update
    with only the diff; (a CNAME-vs-other-type conflict is modeled as a
    separate existing record at the same name with a different type,
    which the caller must resolve by choosing which record wins).
    """
    if not existing_records:
        return {"action": "create", "body": desired}

    current = existing_records[0]
    diff = {}
    for field in ("content", "ttl", "proxied"):
        if field in desired and current.get(field) != desired[field]:
            diff[field] = desired[field]

    if not diff:
        return {"action": "noop", "id": current["id"]}

    return {"action": "update", "id": current["id"], "body": diff}


def resolve_via_dns(domain, record_type):
    """Look up what a public resolver actually returns right now, using
    dnspython. Informational only, used before deciding what to touch.
    """
    import dns.resolver

    try:
        answer = dns.resolver.resolve(domain, record_type)
        return [str(rdata) for rdata in answer]
    except dns.resolver.NXDOMAIN:
        return []
    except dns.resolver.NoAnswer:
        return []


def list_existing_records(name, record_type):
    """List DNS records at the exact name and type through the Cloudflare API."""
    import requests

    headers = {"Authorization": f"Bearer {CLOUDFLARE_API_TOKEN}"}
    params = {"type": record_type, "name.exact": name}
    r = requests.get(
        f"{CF_API}/zones/{CLOUDFLARE_ZONE_ID}/dns_records",
        headers=headers, params=params, timeout=30,
    )
    r.raise_for_status()
    return [
        {
            "id": rec["id"],
            "name": rec["name"],
            "type": rec["type"],
            "content": rec["content"],
            "ttl": rec.get("ttl", 1),
            "proxied": rec.get("proxied", False),
        }
        for rec in r.json()["result"]
    ]


def create_record(body):
    import requests

    headers = {"Authorization": f"Bearer {CLOUDFLARE_API_TOKEN}"}
    if DRY_RUN:
        log.info("[dry run] would create record %s", body)
        return
    r = requests.post(
        f"{CF_API}/zones/{CLOUDFLARE_ZONE_ID}/dns_records",
        headers=headers, json=body, timeout=30,
    )
    r.raise_for_status()
    log.info("Created record %s", body)


def update_record(record_id, body):
    import requests

    headers = {"Authorization": f"Bearer {CLOUDFLARE_API_TOKEN}"}
    if DRY_RUN:
        log.info("[dry run] would update record %s with %s", record_id, body)
        return
    r = requests.patch(
        f"{CF_API}/zones/{CLOUDFLARE_ZONE_ID}/dns_records/{record_id}",
        headers=headers, json=body, timeout=30,
    )
    r.raise_for_status()
    log.info("Updated record %s with %s", record_id, body)


def run():
    if not (CLOUDFLARE_API_TOKEN and CLOUDFLARE_ZONE_ID):
        log.warning("No Cloudflare credentials set. Nothing to reconcile for %s.", DNS_DOMAIN)
        return

    desired = {
        "name": DNS_DOMAIN,
        "type": DNS_RECORD_TYPE,
        "content": DNS_RECORD_CONTENT,
        "ttl": DNS_RECORD_TTL,
        "proxied": DNS_RECORD_PROXIED,
    }

    existing = list_existing_records(DNS_DOMAIN, DNS_RECORD_TYPE)
    plan = plan_dns_write(existing, desired)

    if plan["action"] == "noop":
        log.info("%s already matches the desired state. Nothing to do.", DNS_DOMAIN)
        return

    if plan["action"] == "create":
        log.info("%s has no existing %s record. Creating.", DNS_DOMAIN, DNS_RECORD_TYPE)
        create_record(plan["body"])
        return

    log.info("%s exists but differs. Updating in place instead of duplicating.", DNS_DOMAIN)
    update_record(plan["id"], plan["body"])


if __name__ == "__main__":
    run()

File: duplicate-record-write-conflict/node/duplicate-record-write-conflict.js
/**
 * List DNS records before writing, then create or update instead of
 * blindly POSTing a duplicate. Repairs the write through Cloudflare.
 *
 * Safe by default. Set DRY_RUN=false to let it write.
 *
 * Env vars:
 *   DNS_DOMAIN               the name to check, e.g. "app.example.com"
 *   DNS_RECORD_TYPE          record type to check, default "A"
 *   DNS_RECORD_CONTENT       desired content, e.g. an IP for an A record
 *   DNS_RECORD_TTL           desired TTL, default 300
 *   DNS_RECORD_PROXIED       "true" or "false", default "false"
 *   CLOUDFLARE_API_TOKEN     Cloudflare API token (only needed for repair)
 *   CLOUDFLARE_ZONE_ID       Cloudflare zone id (only needed for repair)
 *   DRY_RUN                  default "true"; set to "false" to actually write
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "app.example.com";
const DNS_RECORD_TYPE = process.env.DNS_RECORD_TYPE || "A";
const DNS_RECORD_CONTENT = process.env.DNS_RECORD_CONTENT || "203.0.113.10";
const DNS_RECORD_TTL = Number(process.env.DNS_RECORD_TTL || 300);
const DNS_RECORD_PROXIED = (process.env.DNS_RECORD_PROXIED || "false").toLowerCase() === "true";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const CF_API = "https://api.cloudflare.com/client/v4";

export function planDnsWrite(existingRecords, desired) {
  // Pure decision function. No I/O.
  // existingRecords: array of { id, name, type, content, ttl, proxied }
  //                  already at (desired.name, desired.type)
  // desired: { name, type, content, ttl, proxied }
  //
  // Returns one of:
  //   { action: "create", body: desired }
  //   { action: "noop", id }
  //   { action: "update", id, body: { changed fields only } }
  //
  // Pure decision logic, no I/O: given zero existing records -> create;
  // given one existing record identical to desired -> noop; given one
  // existing record differing in content/ttl/proxied -> update with only
  // the diff. (A CNAME-vs-other-type conflict is modeled as a separate
  // existing record at the same name with a different type, which the
  // caller must resolve by choosing which record wins.)
  if (!existingRecords.length) {
    return { action: "create", body: desired };
  }

  const current = existingRecords[0];
  const diff = {};
  for (const field of ["content", "ttl", "proxied"]) {
    if (field in desired && current[field] !== desired[field]) {
      diff[field] = desired[field];
    }
  }

  if (Object.keys(diff).length === 0) {
    return { action: "noop", id: current.id };
  }

  return { action: "update", id: current.id, body: diff };
}

async function resolveViaDns(domain, recordType) {
  // Look up what a public resolver actually returns right now, using the
  // built-in dns module. Informational only, used before deciding what to
  // touch.
  const dns = await import("node:dns");
  const { promises: dnsPromises } = dns;
  try {
    if (recordType === "CNAME") return await dnsPromises.resolveCname(domain);
    return await dnsPromises.resolve4(domain);
  } catch (err) {
    if (err.code === "ENODATA" || err.code === "ENOTFOUND") return [];
    throw err;
  }
}

async function listExistingRecords(name, recordType) {
  const headers = { Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}` };
  const params = new URLSearchParams({ type: recordType, "name.exact": name });
  const res = await fetch(
    `${CF_API}/zones/${CLOUDFLARE_ZONE_ID}/dns_records?${params.toString()}`,
    { headers },
  );
  if (!res.ok) throw new Error(`Cloudflare list returned ${res.status}`);
  const body = await res.json();
  return body.result.map((rec) => ({
    id: rec.id,
    name: rec.name,
    type: rec.type,
    content: rec.content,
    ttl: rec.ttl ?? 1,
    proxied: rec.proxied ?? false,
  }));
}

async function createRecord(body) {
  const headers = {
    Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}`,
    "Content-Type": "application/json",
  };
  if (DRY_RUN) {
    console.log(`[dry run] would create record ${JSON.stringify(body)}`);
    return;
  }
  const res = await fetch(`${CF_API}/zones/${CLOUDFLARE_ZONE_ID}/dns_records`, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Cloudflare create returned ${res.status}`);
  console.log(`Created record ${JSON.stringify(body)}`);
}

async function updateRecord(recordId, body) {
  const headers = {
    Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}`,
    "Content-Type": "application/json",
  };
  if (DRY_RUN) {
    console.log(`[dry run] would update record ${recordId} with ${JSON.stringify(body)}`);
    return;
  }
  const res = await fetch(`${CF_API}/zones/${CLOUDFLARE_ZONE_ID}/dns_records/${recordId}`, {
    method: "PATCH",
    headers,
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Cloudflare update returned ${res.status}`);
  console.log(`Updated record ${recordId} with ${JSON.stringify(body)}`);
}

async function run() {
  if (!(CLOUDFLARE_API_TOKEN && CLOUDFLARE_ZONE_ID)) {
    console.warn(`No Cloudflare credentials set. Nothing to reconcile for ${DNS_DOMAIN}.`);
    return;
  }

  const desired = {
    name: DNS_DOMAIN,
    type: DNS_RECORD_TYPE,
    content: DNS_RECORD_CONTENT,
    ttl: DNS_RECORD_TTL,
    proxied: DNS_RECORD_PROXIED,
  };

  const existing = await listExistingRecords(DNS_DOMAIN, DNS_RECORD_TYPE);
  const plan = planDnsWrite(existing, desired);

  if (plan.action === "noop") {
    console.log(`${DNS_DOMAIN} already matches the desired state. Nothing to do.`);
    return;
  }

  if (plan.action === "create") {
    console.log(`${DNS_DOMAIN} has no existing ${DNS_RECORD_TYPE} record. Creating.`);
    await createRecord(plan.body);
    return;
  }

  console.log(`${DNS_DOMAIN} exists but differs. Updating in place instead of duplicating.`);
  await updateRecord(plan.id, plan.body);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
